### DISCUSSION TITLE
Launch a package without logging in

### DISCUSSION THREAD

**Discussion by mtchuenten** [os-js/OS.js#834] [Q&A] [answered]
Hello all,

In my OS.js instance, I activated the login interface such that a user must log in before they can launch packages from the Menu button.
My question is: How can a user launch a package directly from the login page (Note that you don't have access to the Menu button here, where all available packages are listed)? 

For instance, you may want to reset your password because you've forgotten it and launch a Reset Password package from the login page to serve that purpose. You may want to launch other similar packages without being logged in, i.e., without having access to the Menu button.

Regards.

**Comment by andersevenrud** [comment, marked as answer]
Hi.

Only a few services are available when the login screen is visible. You can change this behavior by modifying the service registration, but if you want things like resetting of passwords etc you can create a custom login adapter.

No need to launch applications for this. And still allows for creating UIs etc.

**Reply by mtchuenten** [reply]
Thank you for the feedback @andersevenrud. I was able to create a custom login adapter as shown below:

<details>
  <summary>Pasted source code example</summary>

```js
import { Login } from "@osjs/client";
import logo from "../logo.png";

export default class MyCustomLogin extends Login {
  render() {
    // Set a custom class name
    this.$container.className = "my-custom-login";

    // Add your HTML content
    this.$container.innerHTML = `
        <form action="" id="loginForm" autocomplete="off">
        <div class="osjs-login-logo" data-position="top" style="background-image: url(${logo});"></div>
          <div>
            <input type="text" placeholder="username" id="username" />
            <input type="password" placeholder="password" id="password" />
            <button type="submit" class="pointer">Login</button>
          </div>
          <div class="forgot-pass"><a href="javascript: void(0);" id="forgetPassBtn">Forgot password</a></div>
          <div class="copyright">Copyright @ 2023 Lemunz HMS.</div>
        </form>
    `;

    document.querySelector("#loginForm").addEventListener("submit", (e) => {
      e.preventDefault();
      this.loginUser();
    });

    document.querySelector("#forgetPassBtn").addEventListener("click", () => {
      console.log("handle password forgotten");
    });
  }

  async loginUser() {
    const username = document.querySelector("#username").value;
    const password = document.querySelector("#password").value;

    if (!username || !password) {
      throw new Error("Please provide a username and a password.");
    } else {
      const response = await fetch(
        "https://jsonplaceholder.typicode.com/posts",
        {
          method: "POST",
          body: JSON.stringify({
            title: username,
            body: password,
            userId: 1,
          }),
          headers: {
            "Content-type": "application/json; charset=UTF-8",
          },
        }
      );

      if (response) {
        const json = await response.json();
        console.log(json);
        localStorage.setItem("user", JSON.stringify(json));
        return json;
      } else {
        console.log("login failed");
        throw new Error("Could not log in");
      }
    }
  }

  async logoutUser() {
    console.log("logout");
    localStorage.removeItem("user");
    return true;
  }
}
 
```

</details>

This gives me full control over the login interface, and I can even add a _Forgot password_ feature. My challenge now is to load/redirect to the home page when authentication is successful: `return json;` doesn't help in this regard. How do I achieve that? I also anticipate facing the same issue when logging out.

My adapter is registered in _src/client/index.js_ as shown below:

```
import MyCustomLogin from "./custom-login.js";
...
osjs.register(AuthServiceProvider, {
    before: true,
    args: {
      login: (core, options) => new MyCustomLogin(core, options),
    },
});
...
```

**Reply by andersevenrud** [reply]
It seems like you have both a custom UI **and** a custom auth procedure.

In this case you can make an "Authentication" adapter (HTTP stuff), and a "Login" adapter (UI).

You can communicate with the authentication in your login with events.

1. https://manual.os-js.org/guide/login/#replace
2. https://manual.os-js.org/tutorial/auth/#custom-authentication-adapter

This should solve your issues, because in the auth you *return* the result from your service, and in the login you hook into events to handle the results so you can update your UI states, etc.

**Reply by mtchuenten** [reply]
Hi @andersevenrud,

Thanks a lot. Based on your comments, I could resolve the issue by triggering a "login:post" event from my custom UI. What I'm trying to achieve now is display custom error messages when login fails. 

In my custom authentication adapter, I have this:
```
...
async login(values) {
    // You can transform the form values from login here if you want
    const username = values.username;
    const password = values.password;
    console.log(values);

    if (!username || !password) {
      throw new Error("Please provide a username and a password.");
...
```
And in my custom login UI, I have this
```
this.on("login:error", (err) =>
      console.error("An error occurred while logging in", err);
```
However, _err_ always resolves to _Login failed_, when I expected it to vary based on what was thrown from my custom authentication adapter (for example, I expected it to resolve to _"Please provide a username and a password."_ when no username or password is provided). 
How then do I pass the error message from the authentication adapter back to my custom login adapter?

**Reply by andersevenrud** [reply]
The exception is not forwarded in the login module. Feel free to PR that.

**Reply by andersevenrud** [reply]
I just published `@osjs/client@3.10.0` [^1] where signals were updated [^2] [^3].

You now have access to the exception:

```js
this.on("login:error", (message, exception) => {})
```

[^1]: https://github.com/os-js/osjs-client/commit/f4343b818274a2823ea8e28b0fbb816b2e9c23e9
[^2]: https://github.com/os-js/osjs-client/commit/f3d8acf6d3076ab96be77a7dbca42f426aac3bf8
[^3]: https://github.com/os-js/manual.os-js.org/commit/7dc7e75cf7b566a5c956232552117c2109aee09c

**Reply by mtchuenten** [reply]
Awesome! 

Thanks a lot man, it works like a charm.

**Reply by andersevenrud** [reply]
Cool! Glad to hear it all worked out :)
